hack/generator/pkg/astmodel/type_merger.go

```diff
@@ -0,0 +1,172 @@
+/*
+ * Copyright (c) Microsoft Corporation.
+ * Licensed under the MIT license.
+ */
+
+package astmodel
+
+import (
+	"reflect"
+
+	"github.com/pkg/errors"
+)
+
+// TypeMerger is like a visitor for 2 types.
+//
+// Conceptually it takes (via Add) a list of functions of the form:
+//
+//     func ([ctx interface{},] a <: Type, b <: Type) (Type, error)
+//
+// where `a` and `b` can be concrete types that implement the `Type` interface.
+//
+// When `TypeMerger.Merge(Type, Type)` is invoked, it will iterate through the
+// provided functions and invoke the first one that matches the concrete types
+// passed. If none match then the fallback provided to `NewTypeMerger` will be invoked.
+//
+// The `ctx` argument can optionally be used to “smuggle” additional data down the call-chain.
+type TypeMerger struct {
+	mergers  []mergerRegistration
+	fallback MergerFunc
+}
+
+type mergerRegistration struct {
+	left, right reflect.Type
+	merge       MergerFunc
+}
+
+type MergerFunc func(ctx interface{}, left, right Type) (Type, error)
+
+func NewTypeMerger(fallback MergerFunc) TypeMerger {
+	return TypeMerger{fallback: fallback}
+}
+
+var typeInterface reflect.Type = reflect.TypeOf((*Type)(nil)).Elem() // yuck
+var errorInterface reflect.Type = reflect.TypeOf((*error)(nil)).Elem()
+var mergerFuncType reflect.Type = reflect.TypeOf((*MergerFunc)(nil)).Elem()
+
+type validatedMerger struct {
+	merger                    reflect.Value
+	leftArgType, rightArgType reflect.Type
+	needsCtx                  bool
+}
+
+func validateMerger(merger interface{}) validatedMerger {
+	it := reflect.ValueOf(merger)
+	if it.Kind() != reflect.Func {
+		panic("merger must be a function")
+	}
+
+	mergerType := it.Type()
+
+	if mergerType.NumIn() != 2 && mergerType.NumIn() != 3 {
+		panic("merger must take 2 arguments (left Type, right Type) or 3 arguments (ctx X, left Type, right Type)")
+	}
+
+	argOffset := mergerType.NumIn() - 2
+
+	needsCtx := argOffset != 0
+	leftArg := mergerType.In(argOffset + 0)
+	rightArg := mergerType.In(argOffset + 1)
+
+	if !leftArg.AssignableTo(typeInterface) || !rightArg.AssignableTo(typeInterface) {
+		panic("merger must take in types assignable to Type")
+	}
```

Comment:
Perhaps split this test so you can be explicit about whether it's _left_ or _right_ that's wrong?

Reply:
Thought about it a bit, I'm actually going to go in the other direction and unify this message with the previous one, so if you mess up the argument type at all, it tells you what you need to pass.